### PR TITLE
Expense form dialog open changes

### PR DESCRIPTION
**Pull request by shreyanshss7**

### CODE REVIEW

File: fyle_slack_app/fyle/expenses/views.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+
+    @staticmethod
+    def get_expense_fields_mandatory_mapping(expense_fields: List[Dict]) -> Dict:
+        mandatory_fields_mapping = {
```
Comment (abhishek1234321) [inline fyle_slack_app/fyle/expenses/views.py:107]: why check for only these fields for mandatory?

File: fyle_slack_app/fyle/expenses/views.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            'cost_center': {
+                'is_cost_center_available': is_cost_center_available,
+                'is_mandatory': mandatory_fields_mapping['cost_center_id']
+            },
+            'purpose': mandatory_fields_mapping['purpose'],
+            'transaction_date': mandatory_fields_mapping['txn_dt'],
+            'vendor': mandatory_fields_mapping['vendor_id']
+        }
+
+        additional_currency_details = {
```
Comment (abhishek1234321) [inline fyle_slack_app/fyle/expenses/views.py:153]: currency and amount should also be mandatory

File: fyle_slack_app/fyle/expenses/views.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            'home_currency': home_currency
+        }
+
+        add_to_report = 'existing_report'
```
Comment (abhishek1234321) [inline fyle_slack_app/fyle/expenses/views.py:157]: we should default to existing only if we have reports to add to no?

File: fyle_slack_app/fyle/expenses/views.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        add_to_report = decoded_private_metadata.get('add_to_report')
+
+        current_ui_blocks = slack_payload['view']['blocks']
+
+        project = decoded_private_metadata.get('project')
+
+        custom_field_blocks = []
+
+        for block in current_ui_blocks:
+            if 'custom_field' in block['block_id'] or 'additional_field' in block['block_id']:
```
[the remark this answers is not in the document]
Comment (abhishek1234321) [reply fyle_slack_app/fyle/expenses/views.py:196]: how does this help?

File: fyle_slack_app/fyle/expenses/views.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                custom_field_blocks.append(block)
+
+        if len(custom_field_blocks) == 0:
+            custom_field_blocks = None
```
Comment (abhishek1234321) [inline fyle_slack_app/fyle/expenses/views.py:200]: are we showing all custom fields by default? or only after selection of category?

**Review by abhishek1234321 — CHANGES_REQUESTED** [review]

File: fyle_slack_app/fyle/expenses/views.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+
+    def get_expense_fields(self, query_params: Dict) -> Dict:
+        return self.connection.v1.fyler.expense_fields.list(query_params=query_params)
+
+
+    def get_default_expense_fields(self) -> Dict:
+        default_expense_fields_query_params = {
+            'offset': 0,
+            'limit': '20',
```
[the remark this answers is not in the document]
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:26]: hard limit is 100 I guess on platform side, but limit is a mandatory query param

File: fyle_slack_app/fyle/expenses/views.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+        return self.get_expense_fields(default_expense_fields_query_params)
+
+
+    def get_custom_fields_by_category_id(self, category_id: str) -> Dict:
+        custom_fields_query_params = {
+            'offset': 0,
+            'limit': '20',
+            'order': 'created_at.desc',
+            'column_name': 'not_in.(purpose, txn_dt, vendor_id, cost_center_id)',
```
[the remark this answers is not in the document]
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:42]: There are additional fields as well for taxi category which have `is_custom` as `false`

File: fyle_slack_app/fyle/expenses/views.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            'is_enabled': 'eq.{}'.format(True),
+            'is_custom': 'eq.{}'.format(False),
+            'is_mandatory': 'eq.{}'.format(True)
+        }
+
+        return self.get_expense_fields(default_expense_fields_query_params)
+
+
+    def get_custom_fields_by_category_id(self, category_id: str) -> Dict:
+        custom_fields_query_params = {
```
[the remark this answers is not in the document]
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:38]: Custom we'll only get when their category is selected, is there any other case I missed where we fetch custom fields?

File: fyle_slack_app/fyle/expenses/views.py (hunk 3 above)
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:157]: That check will happen where the UI is rendered.

File: fyle_slack_app/fyle/expenses/views.py (hunk 5 above)
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:200]: only after selection of category

File: fyle_slack_app/fyle/expenses/views.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:196]: This helps in dynamic re-rendering of the form, we use the UI's present in slack payload and get the custom fields directly with the key `custom_field` so we don't have to make an API call again while dynamic re-rendering the form

**Review by abhishek1234321 — CHANGES_REQUESTED** [review]

File: fyle_slack_app/fyle/expenses/views.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+            'offset': 0,
+            'limit': '20',
+            'order': 'created_at.desc',
+            'column_name': 'not_in.(purpose, txn_dt, vendor_id, cost_center_id)',
+            'is_enabled': 'eq.{}'.format(True),
+            'category_ids': 'cs.[{}]'.format(int(category_id))
```
[the remark this answers is not in the document]
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:44]: Yes will implement this once caching PR is merged.

File: fyle_slack_app/fyle/expenses/views.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        return self.get_expense_fields(custom_fields_query_params)
+
+
+    def get_categories(self, query_params: Dict) -> Dict:
```
[the remark this answers is not in the document]
Comment (shreyanshss7) [reply fyle_slack_app/fyle/expenses/views.py:50]: Will create a separate PR for this

**Review by abhishek1234321 — APPROVED** [review]

**Pull request closed without merging**